[WIP] max_pool2d without indices optimization [CPU]

heitorschueroff opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#43267 [WIP] max_pool2d without indices optimization [CPU]**

This PR implements a version of max_pool2d that doesn't compute indices when it's not needed. It also makes some optimizations that will be carried over to other pooling functions in future PRs.

## Benchmarking:

#### Tensor Parameters
BATCH = 10
CHANNEL = 16
HEIGHT = 2048
WIDTH = 2048
DTYPE = torch.float32
DEVICE = "cpu"

#### Pooling Parameters
KERNEL_SIZE = 2
STRIDE = None
PADDING = 0
DILATION = 1
CEIL_MODE = False

#### Results (time in ms) (speedup factor)
test_max_pool2d: 118.4793 (1.0)
test_mkldnn_max_pool2d: 360.2836 (3.04)
test_max_pool2d_with_indices: 626.9831 (5.29)

## Discussion

The new implementation is on average 2~3 times faster than mkldnn and 5x faster than with_indices. The original with_indices code only parallelized over batches and channels, so if these numbers were low it wouldn't achieve optimal parallelism.

This algorithm also reduces duplicate comparisons in the case of overlapping kernel windows. For instance, if we change the pooling parameters above to:

KERNEL_SIZE = 4
STRIDE = 1
PADDING = 1
DILATION = 2
CEIL_MODE = True

#### Results (time in ms) (speedup factor)
test_max_pool2d: 136.4228 (1.0)
test_mkldnn_max_pool2d: 608.4158 (4.46)
test_max_pool2d_with_indices: 1,230.1916 (9.02)

There is also an issue with the existing pooling implementations that they use nested at::parallel_for loops and as such only the outer most loop is parallelized since at::parallel_for does not support nesting.

Differential Revision: [D23273406](https://our.internmc.facebook.com/intern/diff/D23273406)

closes https://github.com/pytorch/pytorch/issues/28733

heitorschueroff committed:
Draft version of max_pool2d without indices optimization

[ghstack-poisoned]

heitorschueroff committed:
Update on "[WIP] max_pool2d without indices optimization"


Included benchmark file for reference. Will remove on final PR.

[ghstack-poisoned]

glaringlee commented on aten/src/ATen/native/Pooling.cpp:
```diff
@@ -1,16 +1,177 @@
+          // Compute kernel first and last rows within input bounds
+          const auto kis = ker_offset(oi * SI - PI);
+          const auto kie =
+              KH - ker_offset(((IH - 1) - (oi * SI + (KH - 1) * DI - PI)));
```
I think ((IH - 1) - (oi * SI + (KH - 1) * DI - PI)) won't < 0, right? (shape check ensured this)
ker_offset is not needed I guess. If so, we can make ker_offset inline here to save a function call.

glaringlee commented on aten/src/ATen/native/Pooling.cpp:
```diff
@@ -1,16 +1,177 @@
+
+          // Compute max column-wise for current output row
+          for (int64_t ki = kis + 1; ki < kie; ++ki) {
+            const auto ii = oi * SI + ki * DI - PI;
```
use int64_t instead auto, no need to deduce this and no need const

glaringlee commented on aten/src/ATen/native/Pooling.cpp:
```diff
@@ -1,16 +1,177 @@
+
+  // Because of padding, the kernel index may be off bounds. This computes the
+  // offset to place the kernel index in bounds.
+  const auto ker_offset = [&](const int64_t excess) {
```
no need to deduce, this is always int64_t, no need const as well

glaringlee commented on aten/src/ATen/native/Pooling.cpp:
```diff
@@ -1,16 +1,177 @@
+  at::parallel_for(
+      0, NB * NC * OH, 0, [&](const int64_t begin, const int64_t end) {
+        for (int64_t it = begin; it < end; ++it) {
+          const auto tensor_idx = it / OH;
```
int64_t for this line and below lines till `const auto copy_offset`. We don't need to deduce it and no need to make it const. They are always int64_t

glaringlee commented on aten/src/ATen/native/Pooling.cpp:
```diff
@@ -1,16 +1,177 @@
+          }
+
+          // Compute max for each cell in the current output row
+          const auto out_offset = (tensor_idx * OH + oi) * OW;
```
tensor_idx * OH  == it I think, you can use it, and add comment to clearify what 'it' is. Use int64_t instead as well.

glaringlee commented on aten/src/ATen/native/Pooling.cpp:
```diff
@@ -127,24 +288,26 @@ Tensor max_pool2d(
   }
 #endif
-  auto output_and_indices = at::max_pool2d_with_indices(
+  if (self.requires_grad()) {
```
I think even require grad, we can still use the new impl, anything blocking here?

glaringlee reviewed (changes requested):
This is overall a great algo and easy to expand to 3d. 
Please see my comments.

heitorschueroff committed:
Update on "[WIP] max_pool2d without indices optimization"


Included benchmark file for reference. Will remove on final PR.

[ghstack-poisoned]

heitorschueroff committed:
Update on "[WIP] max_pool2d without indices optimization"


Included benchmark file for reference. Will remove on final PR.

[ghstack-poisoned]

heitorschueroff committed:
Update on "[WIP] max_pool2d without indices optimization"


Included benchmark file for reference. Will remove on final PR.

[ghstack-poisoned]

glaringlee reviewed (changes requested):
Some minor comments. Please rebase the code, and then I will approve it.

heitorschueroff committed:
Update on "max_pool2d without indices optimization [CPU]"


Included benchmark file for reference. Will remove on final PR.

[ghstack-poisoned]

glaringlee reviewed (commented):
@heitorschueroff 
LGTM to me now except the std::max error.
where is your benchmark?

heitorschueroff committed:
Update on "max_pool2d without indices optimization [CPU]"


Included benchmark file for reference. Will remove on final PR.

[ghstack-poisoned]

heitorschueroff committed:
Update on "max_pool2d without indices optimization [CPU]"


Included benchmark file for reference. Will remove on final PR.

[ghstack-poisoned]

heitorschueroff committed:
Update on "max_pool2d without indices optimization [CPU]"


This PR implements a version of max_pool2d that doesn't compute indices when it's not needed. It also makes some optimizations that will be carried over to other pooling functions in future PRs.

## Benchmarking:

#### Tensor Parameters
BATCH = 10
CHANNEL = 16
HEIGHT = 2048
WIDTH = 2048
DTYPE = torch.float32
DEVICE = "cpu"

#### Pooling Parameters
KERNEL_SIZE = 2
STRIDE = None
PADDING = 0
DILATION = 1
CEIL_MODE = False

#### Results (time in ms) (speedup factor)
test_max_pool2d: 110.0176 (1.0)
test_mkldnn_max_pool2d: 378.5602 (3.44)
test_max_pool2d_with_indices: 626.6335 (5.70)

## Discussion

The new implementation is on average 2~4 times faster than mkldnn and >5x faster than with_indices. The original with_indices code only parallelized over batches and channels, so if these numbers were low it wouldn't achieve optimal parallelism.

This algorithm also reduces duplicate comparisons in the case of overlapping kernel windows. For instance, if we change the pooling parameters above to:

KERNEL_SIZE = 4
STRIDE = 2
PADDING = 2
DILATION = 1
CEIL_MODE = True

#### Results (time in ms) (speedup factor)
test_max_pool2d: 136.4228 (1.0)
test_mkldnn_max_pool2d: 608.4158 (4.46)
test_max_pool2d_with_indices: 1,230.1916 (9.02)

[ghstack-poisoned]

heitorschueroff committed:
Update on "max_pool2d without indices optimization [CPU]"


This PR implements a version of max_pool2d that doesn't compute indices when it's not needed. It also makes some optimizations that will be carried over to other pooling functions in future PRs.

## Benchmarking:

#### Tensor Parameters
BATCH = 10
CHANNEL = 16
HEIGHT = 2048
WIDTH = 2048
DTYPE = torch.float32
DEVICE = "cpu"

#### Pooling Parameters
KERNEL_SIZE = 2
STRIDE = None
PADDING = 0
DILATION = 1
CEIL_MODE = False

#### Results (time in ms) (speedup factor)
test_max_pool2d: 119.0151 (1.0)
test_mkldnn_max_pool2d: 287.4994 (2.42)
test_max_pool2d_with_indices: 639.1541 (5.37)

## Discussion

The new implementation is on average 2~4 times faster than mkldnn and >5x faster than with_indices. The original with_indices code only parallelized over batches and channels, so if these numbers were low it wouldn't achieve optimal parallelism.

This algorithm also reduces duplicate comparisons in the case of overlapping kernel windows. For instance, if we change the pooling parameters above to:

KERNEL_SIZE = 4
STRIDE = 2
PADDING = 2
DILATION = 1
CEIL_MODE = True

#### Results (time in ms) (speedup factor)
test_max_pool2d: 136.4228 (1.0)
test_mkldnn_max_pool2d: 608.4158 (4.46)
test_max_pool2d_with_indices: 1,230.1916 (9.02)

Differential Revision: [D23273406](https://our.internmc.facebook.com/intern/diff/D23273406)

closes https://github.com/pytorch/pytorch/issues/28733

[ghstack-poisoned]

glaringlee reviewed (approved):
LGTM, approving. Please rebase and import to phabricator.

heitorschueroff committed:
Update on "max_pool2d without indices optimization [CPU]"


This PR implements a version of max_pool2d that doesn't compute indices when it's not needed. It also makes some optimizations that will be carried over to other pooling functions in future PRs.

## Benchmarking:

#### Tensor Parameters
BATCH = 10
CHANNEL = 16
HEIGHT = 2048
WIDTH = 2048
DTYPE = torch.float32
DEVICE = "cpu"

#### Pooling Parameters
KERNEL_SIZE = 2
STRIDE = None
PADDING = 0
DILATION = 1
CEIL_MODE = False

#### Results (time in ms) (speedup factor)
test_max_pool2d: 119.0151 (1.0)
test_mkldnn_max_pool2d: 287.4994 (2.42)
test_max_pool2d_with_indices: 639.1541 (5.37)

## Discussion

The new implementation is on average 2~4 times faster than mkldnn and >5x faster than with_indices. The original with_indices code only parallelized over batches and channels, so if these numbers were low it wouldn't achieve optimal parallelism.

This algorithm also reduces duplicate comparisons in the case of overlapping kernel windows. For instance, if we change the pooling parameters above to:

KERNEL_SIZE = 4
STRIDE = 2
PADDING = 2
DILATION = 1
CEIL_MODE = True

#### Results (time in ms) (speedup factor)
test_max_pool2d: 136.4228 (1.0)
test_mkldnn_max_pool2d: 608.4158 (4.46)
test_max_pool2d_with_indices: 1,230.1916 (9.02)

Differential Revision: [D23273406](https://our.internmc.facebook.com/intern/diff/D23273406)

closes https://github.com/pytorch/pytorch/issues/28733

[ghstack-poisoned]

heitorschueroff committed:
Update on "max_pool2d without indices optimization [CPU]"


This PR implements a version of max_pool2d that doesn't compute indices when it's not needed. It also makes some optimizations that will be carried over to other pooling functions in future PRs.

## Benchmarking:

#### Tensor Parameters
BATCH = 10
CHANNEL = 16
HEIGHT = 2048
WIDTH = 2048
DTYPE = torch.float32
DEVICE = "cpu"

#### Pooling Parameters
KERNEL_SIZE = 2
STRIDE = None
PADDING = 0
DILATION = 1
CEIL_MODE = False

#### Results (time in ms) (speedup factor)
test_max_pool2d: 119.0151 (1.0)
test_mkldnn_max_pool2d: 287.4994 (2.42)
test_max_pool2d_with_indices: 639.1541 (5.37)

## Discussion

The new implementation is on average 2~4 times faster than mkldnn and >5x faster than with_indices. The original with_indices code only parallelized over batches and channels, so if these numbers were low it wouldn't achieve optimal parallelism.

This algorithm also reduces duplicate comparisons in the case of overlapping kernel windows. For instance, if we change the pooling parameters above to:

KERNEL_SIZE = 4
STRIDE = 2
PADDING = 2
DILATION = 1
CEIL_MODE = True

#### Results (time in ms) (speedup factor)
test_max_pool2d: 136.4228 (1.0)
test_mkldnn_max_pool2d: 608.4158 (4.46)
test_max_pool2d_with_indices: 1,230.1916 (9.02)

Differential Revision: [D23273406](https://our.internmc.facebook.com/intern/diff/D23273406)

closes https://github.com/pytorch/pytorch/issues/28733

[ghstack-poisoned]

gchanan commented on test/test_nn.py:
```diff
@@ -9810,26 +9810,28 @@ def helper(n, c, h, w, kernel_size, stride=None,
         helper(10, 512, 31, 31, 3, stride=2)
         helper(1, 129, 8, 8, 3, stride=2)
 
-    @onlyCUDA
```
why do you think this was onlyCUDA before?  Isn't your test (on CPU) going to run the same thing twice and check it's the same?  That's fine I guess?

xwang233 replied:
The original purpose of this onlyCUDA was to test CUDA implementation with CPU impl as a reference. I would also suggest that we keep onlyCUDA there, otherwise there would be a duplicate CPU-CPU comparison. 

In order to have a "purely CPU" test, we can have a few hard-coded pooling input and results.

gchanan commented on test/test_nn.py:
```diff
@@ -9810,26 +9810,28 @@ def helper(n, c, h, w, kernel_size, stride=None,
+        for requires_grad in (True, False):
+            helper(2, 8, 4, 4, 2, requires_grad)
+            helper(1, 100000, 32, 32, 4, requires_grad)
+            helper(1, 100000, 1, 4, (1, 4), requires_grad)  # test for max_pool1d
```
is this comment wrong?

mruberry commented on aten/src/ATen/native/Pooling.cpp:
```diff
@@ -127,24 +287,28 @@ Tensor max_pool2d(
   }
 #endif
-  auto output_and_indices = at::max_pool2d_with_indices(
+  if (self.requires_grad() || self.device() != at::kCPU) {
```
What's up with the gradient check here? Maybe another TODO?

heitorschueroff replied:
If we require grad then we need to compute indices for backward pass.

mruberry commented on test/test_nn.py:
```diff
@@ -9810,26 +9810,28 @@ def helper(n, c, h, w, kernel_size, stride=None,
-
-            y = pool(x)
-            ref_y = pool(ref_x)
+            pool = torch.nn.MaxPool2d(kernel_size=ks, return_indices=True)
```
Would you elaborate on this change? In particular:

- In the original return_indices was not set and it default to False. 
- Doesn't your change only affect the return_indices = False codepath?

glaringlee commented:
@heitorschueroff 
I put this back to [WIP] since you will add `with_indices` part as well for this PR. feel free to remove [WIP] once you are ready.

glaringlee reviewed (changes requested):
@heitorschueroff
I put this back to [WIP] since you will add with_indices part as well for this PR. feel free to remove [WIP] once you are ready.

mruberry commented:
Would you post your benchmark script? cc @ngimel for perf, too. Maybe a couple more sizes as a sanity check? 

- inception_v3: batch x 64 x 147 x 147
- googlenet: batch x 64 x 112 x 112

cross with params for inception v3 (kernel size 3, stride 2), googlenet (kernel size 3, stride 2, ceil mode True) and ResNet (kernel size 3, stride, padding 1)

Are there tests that the other options to maxpool2d are working correctly? Like padding, ceil mode, stride, and dilation?

heitorschueroff committed:
Update on "[WIP] max_pool2d without indices optimization [CPU]"


This PR implements a version of max_pool2d that doesn't compute indices when it's not needed. It also makes some optimizations that will be carried over to other pooling functions in future PRs.

## Benchmarking:

#### Tensor Parameters
BATCH = 10
CHANNEL = 16
HEIGHT = 2048
WIDTH = 2048
DTYPE = torch.float32
DEVICE = "cpu"

#### Pooling Parameters
KERNEL_SIZE = 2
STRIDE = None
PADDING = 0
DILATION = 1
CEIL_MODE = False

#### Results (time in ms) (speedup factor)
test_max_pool2d: 118.4793 (1.0)
test_mkldnn_max_pool2d: 360.2836 (3.04)
test_max_pool2d_with_indices: 626.9831 (5.29)

## Discussion

The new implementation is on average 2~3 times faster than mkldnn and 5x faster than with_indices. The original with_indices code only parallelized over batches and channels, so if these numbers were low it wouldn't achieve optimal parallelism.

This algorithm also reduces duplicate comparisons in the case of overlapping kernel windows. For instance, if we change the pooling parameters above to:

KERNEL_SIZE = 4
STRIDE = 1
PADDING = 1
DILATION = 2
CEIL_MODE = True

#### Results (time in ms) (speedup factor)
test_max_pool2d: 136.4228 (1.0)
test_mkldnn_max_pool2d: 608.4158 (4.46)
test_max_pool2d_with_indices: 1,230.1916 (9.02)

There is also an issue with the existing pooling implementations that they use nested at::parallel_for loops and as such only the outer most loop is parallelized since at::parallel_for does not support nesting.

Differential Revision: [D23273406](https://our.internmc.facebook.com/intern/diff/D23273406)

closes https://github.com/pytorch/pytorch/issues/28733

[ghstack-poisoned]

heitorschueroff commented on test/test_nn.py:
```diff
@@ -9810,26 +9810,28 @@ def helper(n, c, h, w, kernel_size, stride=None,
-        def helper(n, c, h, w, ks):
-            x = torch.randn(n, c, h, w, device='cuda', dtype=torch.float, requires_grad=True)
+        def helper(n, c, h, w, ks, requires_grad):
+            x = torch.randn(n, c, h, w, device=device, dtype=torch.float, requires_grad=requires_grad)
```
x.to('cpu', copy=True).requires_grad_() is returning None for some reason.

heitorschueroff replied:
or rather calling .grad on the returning Tensor is None

xwang233 reviewed (commented):
Besides your changes to `test_max_pool2d`, there is also another `test_max_pool2d_indices`. Would you mind combine the two tests together, rather than modifying the current `test_max_pool2d` to another duplicates of "test_another_max_pool2d_indices"? Thanks!

https://github.com/pytorch/pytorch/blob/42a53607359af5c84638318d30b939f48c6aab8e/test/test_nn.py#L9868-L9889
